### PR TITLE
feat(nxp/pxp): add Zephyr Support

### PR DESCRIPTION
**Pull request by faxe1008**
### Description of the feature or fix

Reopening of [https://github.com/lvgl/lvgl/pull/6159](https://github.com/lvgl/lvgl/pull/6159) since it got merged and reverted the original PR does not track changes anymore (@nicusorcitu ). 

### Notes
- Update the [Documentation](https://github.com/lvgl/lvgl/tree/master/docs) if needed.
- Add [Examples](https://github.com/lvgl/lvgl/tree/master/examples) if relevant.
- Add [Tests](https://github.com/lvgl/lvgl/blob/master/tests/README.md) if applicable.
- If you added new options to `lv_conf_template.h` run [lv_conf_internal_gen.py](https://github.com/lvgl/lvgl/blob/master/scripts/lv_conf_internal_gen.py) and update [Kconfig](https://github.com/lvgl/lvgl/blob/master/Kconfig).
- Run `scripts/code-format.py` ([astyle](http://astyle.sourceforge.net/install.html) version [v3.4.12](https://github.com/szepeviktor/astyle/releases/tag/v3.4.12) needs to be installed) and follow the [Code Conventions](https://docs.lvgl.io/master/CODING_STYLE.html).
- Mark the Pull request as [Draft](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/changing-the-stage-of-a-pull-request) while you are working on the first version, and mark is as _Ready_ when it's ready for review.
- When changes were requested, [re-request review](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/requesting-a-pull-request-review) to notify the maintainers.
- Help us to review this Pull Request! Anyone can [approve or request changes](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/approving-a-pull-request-with-required-reviews).

### CODE REVIEW

**Comment by nicusorcitu** [comment]
@cristian-stoica

File: src/draw/nxp/pxp/lv_pxp_osa.c
Comment (nicusorcitu) [inline src/draw/nxp/pxp/lv_pxp_osa.c:28]: Those should not be included - they will get over lv_os.h inclusion

File: src/draw/nxp/pxp/lv_pxp_osa.c
Comment (nicusorcitu) [inline src/draw/nxp/pxp/lv_pxp_osa.c]: I understand now, you included the ZEPHYR header just for this `CONFIG_LV_Z_PXP_INTERRUPT_PRIORITY`.

We should not mix any longer this with lv_osal support.

File: src/draw/nxp/pxp/lv_pxp_osa.c
Comment (cristian-stoica) [inline src/draw/nxp/pxp/lv_pxp_osa.c:136]: With PXP=1 and OS_NONE the build fails on MCUXpresso SDK projects. This is a common build configuration for us. The build fails because SDK_OS_FREE_RTOS is defined (by MCUX) but configLIBRARY_MAX_SYSCALL_INTERRUPT_PRIORITY is no longer found.

Since you similarly need CONFIG_LV_Z_PXP_INTERRUPT_PRIORITY for __ZEPHYR__ we could probably use a common abstraction for interrupt priory that will resolve the build for both cases. What do you think we should do about this?

Comment (nicusorcitu) [reply src/draw/nxp/pxp/lv_pxp_osa.c:136]: hi @faxe1008 
Can you please address this issue? Try to not mix anylonger the `SDK_OS_FREE_RTOS` /`__ZEPHYR__` with `LV_USE_OS`.
We can fix by ourself the other problem with multithreading mentioned [here](https://github.com/lvgl/lvgl/pull/6159#discussion_r1593353421) but please fixup this PR before merging.

File: src/draw/nxp/pxp/lv_pxp_osa.c
Comment (faxe1008) [reply src/draw/nxp/pxp/lv_pxp_osa.c]: The includes are needed for the entire interrupt logic (retrieving nodes from DT, hooking up the IRQ).

File: src/draw/nxp/pxp/lv_pxp_osa.c
Comment (faxe1008) [reply src/draw/nxp/pxp/lv_pxp_osa.c:136]: @nicusorcitu I am not quite sure if I understood everything correctly but I ammended the commit with a change that hopefully fixes this.

Comment (nicusorcitu) [reply src/draw/nxp/pxp/lv_pxp_osa.c:136]: Actually it makes it worst because now there is no distinction at all between FreeRTOS and Bare Metal for PXP.

What I men was to have kind of an abstraction os function like "lv_os_interrupt_enable()" or "lv_os_interrupt_disable()" but I figure out that this is basically something platform dependent and not OS dependent.
@cristian-stoica suggested a simpler way to have only an abstracted define in lv osal:
```
#if LV_USE_OS == LV_OS_FREERTOS
  #define LV_OS_INTERRUPT_PRIORITY configLIBRARY_MAX_SYSCALL_INTERRUPT_PRIORITY +1
#elif LV_USE_OS == LV_OS_ZEPHYR
  #define LV_OS_INTERRUPT_PRIORITY CONFIG_LV_Z_PXP_INTERRUPT_PRIORITY
#endif
```
But we still do not get rid of the extra include headers for FreeRTOS or Zephyr, correct?

To keep things simple and conclude on this PR:
1. please revert your last changes from [here](https://github.com/lvgl/lvgl/compare/fd1e2466a5133625483e819ee3bf08738d80b9f2..158a93acf415225cac7c74f0d1a0536a76b793fa)
2. add back the inclusion necessary for freertos irq priority (for freertos)
```
#if defined(SDK_OS_FREE_RTOS)
    #include "FreeRTOS.h"
#endif
```

Comment (faxe1008) [reply src/draw/nxp/pxp/lv_pxp_osa.c:136]: Performed the suggested changes :^)

Comment (nicusorcitu) [reply src/draw/nxp/pxp/lv_pxp_osa.c:136]: I do not see them. Can you please point a link?

**Comment by cristian-stoica** [comment]
-#if defined(SDK_OS_FREE_RTOS)
-    #include "FreeRTOS.h"
-    #include "semphr.h"
+
+#if defined(__ZEPHYR__)
+    #include <zephyr/kernel.h>
+    #include <zephyr/irq.h>
 #endif
 
 I have a problem with the code above. It is not exactly OSAL as it claims and it also breaks our build. There has to be a better way

File: src/draw/nxp/pxp/lv_pxp_osa.c
Comment (faxe1008) [reply src/draw/nxp/pxp/lv_pxp_osa.c:136]: I forgot to push the changes and was out of office for the week. Pushed it right now :^)

**Comment by faxe1008** [comment]
> ```
> 
> -#if defined(SDK_OS_FREE_RTOS)
> -    #include "FreeRTOS.h"
> -    #include "semphr.h"
> +
> +#if defined(__ZEPHYR__)
> +    #include <zephyr/kernel.h>
> +    #include <zephyr/irq.h>
> #endif
> ```
> 
> I have a problem with the code above. It is not exactly OSAL as it claims and it also breaks our build. There has to be a better way

Yeah, these includes are needed for the interrupts. Personally I think that an OSAL is maybe not the right place for an abstraction for this, the OSAL should be kept as thin as possible. Maybe it would be an option somehow provide the user a way to give LVGL a `setup` and `teardown` method for any of the drawing engines so that this can be handled by each OS?

File: src/draw/nxp/pxp/lv_pxp_osa.c
Comment (nicusorcitu) [inline src/draw/nxp/pxp/lv_pxp_osa.c]: The is still one issue:
`lv_thread_sync_signal()` was not ment to work work from ISR.

The implementation from ISR should be using FreeRTOS `vTaskNotifyGiveFromISR()` instead of `xTaskNotifyGive()`.

Comment (nicusorcitu) [reply src/draw/nxp/pxp/lv_pxp_osa.c]: We should add a new osal API in FreeRTOS (and same for Zephyr) like `lv_thread_sync_sygnal_from_isr()`.

Comment (faxe1008) [reply src/draw/nxp/pxp/lv_pxp_osa.c]: I added this method to the OSAL, however I would need help from experts of all other plattforms if I have done it correctly.

Comment (nicusorcitu) [reply src/draw/nxp/pxp/lv_pxp_osa.c]: The API does not look good on FreeRTOS either.
1. You should use `vTaskNotifyGiveFromISR()`
2. From documentation:
```
vTaskNotifyGiveFromISR() will set *pxHigherPriorityTaskWoken to pdTRUE if sending the notification caused a task to unblock, and the unblocked task has a priority higher than the currently running task.

If vTaskNotifyGiveFromISR() sets this value to pdTRUE then a context switch should be requested before the interrupt is exited. See the example below. 
```


```
/* The transmit end interrupt. */
void vTransmitEndISR( void )
{
BaseType_t xHigherPriorityTaskWoken = pdFALSE;

    /* At this point xTaskToNotify should not be NULL as a transmission was
    in progress. */
    configASSERT( xTaskToNotify != NULL );

    /* Notify the task that the transmission is complete. */
    vTaskNotifyGiveIndexedFromISR( xTaskToNotify, 0, &xHigherPriorityTaskWoken );

    /* There are no transmissions in progress, so no tasks to notify. */
    xTaskToNotify = NULL;

    /* If xHigherPriorityTaskWoken is now set to pdTRUE then a context switch
    should be performed to ensure the interrupt returns directly to the highest
    priority task.  The macro used for this purpose is dependent on the port in
    use and may be called portEND_SWITCHING_ISR(). */
    portYIELD_FROM_ISR( xHigherPriorityTaskWoken );
}
```

File: src/osal/lv_freertos.c
Comment (nicusorcitu) [inline src/osal/lv_freertos.c]: Please don't use NULL.
Take care of xHigherPriorityTaskWoken --> I already copy-paste an example on how to use it.

File: src/osal/lv_freertos.c
Comment (nicusorcitu) [inline src/osal/lv_freertos.c]: This API need to be tested with USE_FREERTOS_TASK_NOTIFY(task_notify implementation) or without USE_FREERTOS_TASK_NOTIFY (semaphores implementation).

**Comment by kisvegabor** [comment]
We got a lot request about updating Zephyr to LVGL v9 so I wonder if we can help to make it happen soon. @becseya is interesting Zephyr, so he can help with LVGL related issues. We also have several NXP boards so we can help with PXP as well.

File: src/osal/lv_freertos.c
Comment (faxe1008) [reply src/osal/lv_freertos.c]: Applied the suggestion, also for the other branch. Hopefully correct since I am not an Free RTOS expert :^)

**Comment by kisvegabor** [comment]
>   155 | lv_result_t lv_thread_sync_signal_isr(lv_thread_sync_t * sync)

Please add `LV_UNUSED(sync);`

**Comment by faxe1008** [comment]
The function var had to be renamed to `pxCond` :^)

**Comment by kisvegabor** [comment]
> The function var had to be renamed to pxCond

Which one? :slightly_smiling_face:

**Comment by nicusorcitu** [comment]
image

As a general rule, inside the FreeRTOS files I have used FreeRTOS style while naming variables, functions...
Thanks for keeping it on the same manner.

**Review by nicusorcitu — APPROVED** [review]

**Comment by kisvegabor** [comment]
Please add `lv_thread_sync_signal_isr` to `lv_os_none.c` too.

**Comment by faxe1008** [comment]
@kisvegabor added it to `lv_os_none` :^)

**Review by kisvegabor — DISMISSED** [review]

File: src/draw/nxp/pxp/lv_pxp_osa.c
Comment (liamHowatt) [inline src/draw/nxp/pxp/lv_pxp_osa.c]: Please put this in the static functions section and forward declare it. Maybe also change the name to make it seem less like a platform specific global function definition (unless it is).

File: src/osal/lv_freertos.c
Comment (liamHowatt) [inline src/osal/lv_freertos.c]: If this fails because `lv_thread_sync_wait` has the mutex, will the signal never be sent and a deadlock would occur?

**Review by liamHowatt — COMMENTED** [review]
Nice atomic CAS loop!

In another PR we should correct the implementation of `lv_mutex_lock_isr` for cmsis_rtos2. See [this comment](https://github.com/lvgl/lvgl/pull/6182#pullrequestreview-2084004168).

File: src/draw/nxp/pxp/lv_pxp_osa.c
Comment (faxe1008) [reply src/draw/nxp/pxp/lv_pxp_osa.c]: Ah it kinda is - I do not know how many other RTOS give the possibility to add userdata to an ISR.

File: src/osal/lv_freertos.c
Comment (faxe1008) [reply src/osal/lv_freertos.c]: True - what would be the best solution for this? Not using the mutex at all and disabling interrupts instead so we can not be preempted by another ISR?

Comment (faxe1008) [reply src/osal/lv_freertos.c]: See my last commit, if the change is ok from your side, I'll squash it :^)

File: src/draw/nxp/pxp/lv_pxp_osa.c
Comment (liamHowatt) [reply src/draw/nxp/pxp/lv_pxp_osa.c]: What I mean is the name implies that it's overriding a `weak` definition in the OS like a systick handler that gets called by the OS, but it's really just a local function.

Also, does `PXP_IRQHandler` need to be global? Does it get called by anything external?

File: src/osal/lv_freertos.c
Comment (liamHowatt) [reply src/osal/lv_freertos.c]: Yes, I think that should work, nice. Is the CAS loop no longer needed now?

File: src/draw/nxp/pxp/lv_pxp_osa.c
Comment (faxe1008) [reply src/draw/nxp/pxp/lv_pxp_osa.c]: Ah, no its really just a wrapper to accept the `void *` that the Zephyr impl provides. 
`PXP_IRQHandler` does not need to be global - but that I did not touch. Should I make it a static one as well?

File: src/osal/lv_freertos.c
Comment (faxe1008) [reply src/osal/lv_freertos.c]: Correct, I removed the CAS loop.

Comment (liamHowatt) [reply src/osal/lv_freertos.c]: Did you? :wink:

Comment (faxe1008) [reply src/osal/lv_freertos.c]: Now I did :'D

Comment (liamHowatt) [reply src/osal/lv_freertos.c]: Oh, you pushed right after I sent that :laughing:

File: src/draw/nxp/pxp/lv_pxp_osa.c
Comment (liamHowatt) [reply src/draw/nxp/pxp/lv_pxp_osa.c]: Yeah, it's best to not add to the global namespace if it can be avoided. For now, just make `PXP_IRQHandler` static and move this heading above `PXP_IRQHandler` please, and it all looks good :slightly_smiling_face:

```c
/**********************
 *   STATIC FUNCTIONS
 **********************/
```

File: src/osal/lv_freertos.c
Comment (kisvegabor) [reply src/osal/lv_freertos.c]: @liamHowatt please approve if all looks good.

**Review by kisvegabor — DISMISSED** [review]

File: src/draw/nxp/pxp/lv_pxp_osa.c
Comment (faxe1008) [reply src/draw/nxp/pxp/lv_pxp_osa.c]: Performed the change and renamed the handler to match the other non namespace scoped functions.

**Review by liamHowatt — DISMISSED** [review]
Thank you!

File: src/draw/nxp/pxp/lv_pxp_osa.c
Comment (nicusorcitu) [inline src/draw/nxp/pxp/lv_pxp_osa.c]: Please do not remove the PXP_IRQHandler(). @faxe1008

Comment (faxe1008) [reply src/draw/nxp/pxp/lv_pxp_osa.c]: Within LVGL this is not used outside of the osa file - is this something that needs to be callable from the outside?

Comment (nicusorcitu) [reply src/draw/nxp/pxp/lv_pxp_osa.c]: `PXP_IRQHandler()` is what the name suggest - the handler of the PXP IRQ.

It is similar to what you already added `PXP_ZephyrIRQHandler` except that in case of our platform (w/ or w/o FreeRTOS) the PXP handler is mapped in the vector table outside of the lvgl code.

Comment (liamHowatt) [reply src/draw/nxp/pxp/lv_pxp_osa.c]: Okay, so that function _is_ called from outside? I wasn't clear on that.

Even though it wasn't perfect, I approved yesterday because it was harmless and close enough. But the naming is still confusing because `PXP_IRQHandler` is actually called from outside and the name is important, because it maps to something in the symbol table as you say, while `PXP_ZephyrIRQHandler` is not special yet has a name that makes it seem like it maps to something in the symbol table.

Comment (nicusorcitu) [reply src/draw/nxp/pxp/lv_pxp_osa.c]: In this case we can rename the `PXP_ZephyrIRQHandler` as this is a static function.
But `PXP_IRQHandler` need to stay in place with exactly the same name.

Thanks.

Comment (faxe1008) [reply src/draw/nxp/pxp/lv_pxp_osa.c]: Reverted this part of the change moved the zephyr handler to the static section.

File: src/draw/nxp/pxp/lv_pxp_osa.c
Comment (faxe1008) [inline src/draw/nxp/pxp/lv_pxp_osa.c]: Gotta remove the semicolon :^)

File: src/draw/nxp/pxp/lv_pxp_osa.c
Comment (nicusorcitu) [inline src/draw/nxp/pxp/lv_pxp_osa.c:143]: Please move this line before NVIC_EnableIRQ() like before.

File: src/draw/nxp/pxp/lv_pxp_osa.c
Comment (nicusorcitu) [inline src/draw/nxp/pxp/lv_pxp_osa.c:79]: As this file no longer refer to FreeRTOS only, please rename this variable to "pxp_idle". Has nothing to do directly with your changes but please include this change as well - if not big deal. Thanks.

File: src/draw/nxp/pxp/lv_pxp_osa.c
Comment (nicusorcitu) [inline src/draw/nxp/pxp/lv_pxp_osa.c:116]: Why do you ever need the static `_pxp_zephyr_irq_handler`? You can use directly `PXP_IRQHandler()` for zephyr as well.

Comment (faxe1008) [reply src/draw/nxp/pxp/lv_pxp_osa.c:116]: I can not use it, the `IRQ_CONNECT` macro needs a function with a `(void)(void*)` signature:
[https://docs.zephyrproject.org/apidoc/latest/group__isr__apis.html#ga131739d1faf501a15590053817aba984](https://docs.zephyrproject.org/apidoc/latest/group__isr__apis.html#ga131739d1faf501a15590053817aba984)

**Comment by kisvegabor** [comment]
Hi,

What is missing to finalize this PR and for LVGL v9 Zephyr support in general? I'd like to be sure all are added before releasing v9.2.

**Comment by faxe1008** [comment]
@kisvegabor at least from my side there are no more blockers to this PR. I guess some testing from the other RTOSes might be helpful to avoid regressions.

As for the requirements of updating zephyr to v9.2 the only thing left for me is to figure out what to do about STM32 DMA2D support, as we have added it in 8.4 but I am unsure if it can be used in 9.1.

**Comment by nicusorcitu** [comment]
I have already approved, no more issues from my side.

**Review by nicusorcitu — APPROVED** [review]

**Review by liamHowatt — APPROVED** [review]

**Review by kisvegabor — APPROVED** [review]

**Comment by kisvegabor** [comment]
Great! Is the missing DMA2D is blocking to have LVGL v9.2 in Zephyr? We got closer to ST, so now we can add DMA2D support but it might take some time.

**Comment by faxe1008** [comment]
> Great! Is the missing DMA2D is blocking to have LVGL v9.2 in Zephyr? We got closer to ST, so now we can add DMA2D support but it might take some time.

People using the new version on STM32 boards will experience a performance hit, but imo that should be accetable to some degree. Btw. I started drafting this implementation over here:
https://github.com/faxe1008/lvgl/commit/6497837e396fb2b91804158f7e1aed667a69f0a6
There is still a lot of stuff missing, I used the old implementation as a kind of guide line and some datasheets, but I am by no means an expert on ChromArt.

**Comment by kisvegabor** [comment]
@liamHowatt is also looking into it now. I'd like to have at least something basic in v9.2.

**Pull request merged**
